### ISSUE TITLE
Revise type hierarchy

### ISSUE THREAD

**Issue by timueh** [timueh/PolyChaos.jl#22]
The current type hierarchy is alright-ish, but not the cleanest-code version we can think of. It'd be cleaner to introduce abstract types and introduce specific types, e.g., for canonical uncertainties.

Something like this

```julia
abstract type AbstractMeasure end
abstract type AbstractCanonicalMeasure <: AbstractMeasure end

struct GaussMeasure <: AbstractCanonicalMeasure
    w::Function
    dom::Tuple{Real, Real}
    function GaussMeasure()
        new(gauss_density, (-Inf, Inf))
    end
end

# add other canonical measures below
```

**Comment by timueh** [comment]
We use the branch [`ReviseTypes`](https://github.com/timueh/PolyChaos.jl/tree/ReviseTypes) to work on this.

**Comment by timueh** [comment]
@FrederikZa shouldn't this (and the other canonical ortho polys)

https://github.com/timueh/PolyChaos.jl/blob/e394ed72e095b7d290a2998ee1e895e02150f056/src/types.jl#L198

read

```julia
struct LegendreOrthoPoly <: AbstractCanonicalOrthoPoly 
```

**Comment by timueh** [comment]
Yay, we're making progress!

@FrederikZa: I checked the scalar products. They should be fine. The next step is to check out what tests fail on travis-ci and fix those:

https://travis-ci.org/timueh/PolyChaos.jl

The next test file to fix is `recurrence_coefficients.jl`

**Comment by timueh** [comment]
adjusted `src/polynomial_chaos.jl` to comply with new types via https://github.com/timueh/PolyChaos.jl/commit/e24922f9f48edaad089dcee363184c008c881d31

**Comment by timueh** [comment]
Issue closed with [version 0.2.0](https://github.com/timueh/PolyChaos.jl/releases/tag/v0.2.0)
